`mir_slice_value` and `mir_slice_range_value` (#1948)
GaloisInc/saw-script: doc/manual/manual.md

* `mir_tuple_value : [MIRValue] -> MIRValue` construct a tuple with the given
  list of values as elements.

> ### MIR slices

Author: Documentation

* `mir_tuple_value : [MIRValue] -> MIRValue` construct a tuple with the given
  list of values as elements.

### MIR slices

Slices are a unique form of compound type that is currently only used during
MIR verification. Unlike other forms of compound values, such as arrays, it is
not possible to directly construct a slice. Instead, one must take a slice of
an existing reference value that points to the thing being sliced. The
following commands are used to construct slices:

* `mir_slice_value : MIRValue -> MIRValue`: the SAWScript expression

>   `mir_slice_value base` is equivalent to the Rust expression `base[..]`,

Reviewer: ```suggestion
  `mir_slice_value base` is equivalent to the Rust expression `&base[..]`,
```
Clarify that the result has type `&[T]` (or `&mut [T]`, I suppose) rather than `[T]`

>   i.e., a slice of the entirety of `base`. `base` must be a reference to
>   an array value.

Reviewer: ```suggestion
  i.e., a slice of the entirety of `base`. `base` must be a reference to
  an array (`&[T; N]` or `&mut [T; N]`), not an array itself.
```
More clarification.  You could instead say `&(*base)[..]` above, but that's a bit verbose IMO

* `mir_slice_range_value : MIRValue -> Int -> Int -> MIRValue`: the SAWScript
  expression `mir_slice_range_value base start end` is equivalent to the Rust

>   expression `base[start..end]`, i.e., a slice over a part of `base` which

Reviewer: ```suggestion
  expression `&base[start..end]`, i.e., a slice over a part of `base` which
```
Same as above

  ranges from `start` to `end`. `base` must be a reference to an array value,
  and `start` and `end` are assumed to be zero-indexed. `start` must not exceed
  `end`, and `end` must be strictly less than the length of the array that

>   `base` points to.

Reviewer: In Rust, `start..end` is a half-open range, so `end == len` is valid.  For example:
```Rust
let arr = [1, 2, 3, 4, 5];
let slice = &arr[0..5];
```
`slice` is a 5-element slice containing all the elements of `arr`.

Author: Another occurrence of my earlier typo leaking out. I've corrected this to say "`end` must not exceed the length of the array".

As an example of how to use these functions, consider this Rust function, which
accepts an arbitrary slice as an argument:

~~~~ .rs
pub fn f(s: &[u32]) -> u32 {
    s[0] + s[1]
}
~~~~

We can write a specification that passes a slice to the array `[1, 2, 3, 4, 5]`
as an argument to `f`:

~~~~
let f_spec_1 = do {
  a <- mir_alloc (mir_array 5 mir_u32);
  mir_points_to a (mir_term {{ [1, 2, 3, 4, 5] : [5][32] }});

  mir_execute_func [mir_slice_value a];

  mir_return (mir_term {{ 3 : [32] }});
};
~~~~

Alternatively, we can write a specification that passes a part of this array
which ranges from second element to the fourth.

~~~~
let f_spec_2 = do {
  a <- mir_alloc (mir_array 5 mir_u32);
  mir_points_to a (mir_term {{ [1, 2, 3, 4, 5] : [5][32] }});

  mir_execute_func [mir_slice_range_value a 1 3];

  mir_return (mir_term {{ 5 : [32] }});
};

> ~~~~

Reviewer: >from second element to the fourth

Should note that this is a half-open range, so the resulting slice has length 2.
